File: cslavb/Csla/Linq/IIndex.vb
Imports System
Imports System.Collections.Generic
Imports System.Reflection

Namespace Linq
  ''' <summary>
  ''' Interface that determines functionality of an index
  ''' </summary>
  Public Interface IIndex(Of T)
    Inherits ICollection(Of T)
    ''' <summary>
    ''' Field this index is indexing on
    ''' </summary>
    ReadOnly Property IndexField() As PropertyInfo
    ''' <summary>
    ''' Iterator that returns objects where there is a match based on the passed item
    ''' </summary>
    Function WhereEqual(ByVal item As T) As IEnumerable(Of T)
    ''' <summary>
    ''' Iterator that returns objects based on the expression and the item hashcode passed in
    ''' </summary>
    Function WhereEqual(ByVal pivotVal As Object, ByVal expr As Func(Of T, Boolean)) As IEnumerable(Of T)
    ''' <summary>
    ''' Reindex an item in this index
    ''' </summary>
    Sub ReIndex(ByVal item As T)
    ''' <summary>
    ''' Determine whether the given index is loaded or not
    ''' </summary>
    ReadOnly Property Loaded() As Boolean
    ''' <summary>
    ''' Set the index as not loaded anymore
    ''' </summary>
    Sub InvalidateIndex()
    ''' <summary>
    ''' Set the index as as loaded
    ''' </summary>
    Sub LoadComplete()
    ''' <summary>
    ''' Determine the index mode (always, ondemand, never)
    ''' </summary>
    Property IndexMode() As IndexModeEnum
  End Interface
End Namespace

File: cslacs/Csla/Linq/IIndex.cs
using System;
using System.Collections.Generic;
using System.Reflection;

namespace Csla.Linq
{
  /// <summary>
  /// Interface that determines functionality of an index
  /// </summary>
  public interface IIndex<T> : ICollection<T>
  {
    /// <summary>
    /// Field this index is indexing on
    /// </summary>
    PropertyInfo IndexField { get; }
    /// <summary>
    /// Iterator that returns objects where there is a match based on the passed item
    /// </summary>
    IEnumerable<T> WhereEqual(T item);
    /// <summary>
    /// Iterator that returns objects based on the expression and the value passed in
    /// </summary>
    IEnumerable<T> WhereEqual(object pivotVal, Func<T, bool> expr);
    /// <summary>
    /// Reindex an item in this index
    /// </summary>
    void ReIndex(T item);
    /// <summary>
    /// Determine whether the given index is loaded or not
    /// </summary>
    bool Loaded { get; }
    /// <summary>
    /// Set the index as not loaded anymore
    /// </summary>
    void InvalidateIndex();
    /// <summary>
    /// Set the index as as loaded
    /// </summary>
    void LoadComplete();
    /// <summary>
    /// Determine the index mode (always, ondemand, never)
    /// </summary>
    IndexModeEnum IndexMode { get; set; }
  }
}
